### ISSUE TITLE
Consider single-responsibility-principle for AsNoTracking() with fix-ups

### ISSUE THREAD

**Issue by alexk8** [aspnet/EntityFrameworkCore#11227]
Nowadays (efcore 2.0 and 2.1-pre1), AsNoTracking() does not do fix-ups for reference-properties (practically) and, as far as I know, this is considered as a bug and going to be fixed in 2.1 (issue #10603).
Some time ago it did, and some people complained that they actually do not want it happen (see docs comments at the bottom: https://docs.microsoft.com/en-us/ef/core/querying/tracking).
So I think, AsNoTracking() either should not care about Fix-ups and delegate it to other method like SkipFixups(), or should be configurable - AsNoTracking(bool fixups=true), because most scenarios need fix-ups, but there are obviously exist some where they are extremly undesirable.

**Comment by ajcvickers** [comment]
@alexk8 Some questions to clarify what you are asking. First, the current behavior is:
* An entity returned from a tracking query is fixed up for both sides of all relationships to any other tracked entity.
* An entity returned from a no-tracking query is fixed up for both sides of only the relationships explicitly specified in Include.

For example, in this model:
```C#
public class Blog
{
    public int Id { get; set; }
    public ICollection<Post> Posts { get; set; }
}

public class Post
{
    public int Id { get; set; }
    public Blog Blog { get; set; }

    public Post NextPost { get; set; }
    public Post PreviousPost { get; set; }
}
```
There are two relationships:
* A one-to-many between Blog and Post, with navigation properties on both sides: Blog.Posts and Post.Blog
* A one-to-one self-referencing between two Posts, with navigation properties one both sides: Post.NextPost and Post.PreviousPost.

A query like this:
```C#
context.Set<Post>().Include(p => p.Blogs).ToList()
```
when run as a no-tracking query will bring in all Posts and all related Blogs, and will fixup both sides of the relationship specified by Include. That is the Blog.Posts collection, and the inverse Post.Blog references. However, neither side of the other relationship is fixed up because it wasn't specified in Include. On the other hand, when run as a tracking query, the one-to-one relationship will get fixed up because when tracking EF fixes up all relationships between tracked entities.

So that's the current behavior. Based on this, I _think_ what you are asking is to be able to do something like this:
```C#
context.Set<Post>().Include(p => p.Blogs).ToList()
```
on a no-tracking query and have only one side of the relationship fixed up. In this case, only Post.Blogs, and not the inverse Blog.Post properties. Is this correct? If so, given that the same entities are returned either way, what is the advantage of not fixing up the inverse navigation property of the relationship?

**Comment by alexk8** [comment]
@ajcvickers, unfortunately, that's not what I meant.

With tracking behaviour entities are fixed up with ALL entities that returned with query AND already existing in context.
With non-tracking behaviour, now entities are fixed up with ONLY entities that returned within query.
I think that: 
1. non-tracking entities should have fixed-up references to tracking ones, which already exist in context (in most scenarios). At least one-way, maybe, I didnt think about it 
2. sometimes this should not happen.

So, to sum up, I think that "tracking" and "fixing up" are different concerns.
"No-tracking" means only that entities should not go to DbContext state, but they can have links to tracking ones.

Sorry if I didnt explain clearly

**Comment by alexk8** [comment]
Consider this:
```
var list = ctx.BigTable
   .Include(x=>x.RefToSmallTable1.Something1)
   .Include(x=>x.RefToSmallTable2.Something2)
   .Include(x=>x.RefToSmallTable3.Something3)
   .AsNoTracking().ToList();
foreach(var x in list) Console.WriteLine(x.RefToSmallTable1.Something1.xxx)
```   
Which makes lots of joins, that are practically unacceptable.
In contrast with:   
```
ctx.SmallTable1.Include(x=>x.Something1).Load();
ctx.SmallTable2.Include(x=>x.Something2).Load();
ctx.SmallTable3.Include(x=>x.Something3).Load();
var list = ctx.BigTable.AsNoTracking().ToList();
foreach(var x in list) Console.WriteLine(x.RefToSmallTable1.Something1.xxx)
```
Which could use no joins and load far less columns, but now doesnt work

**Comment by ajcvickers** [comment]
@alexk8 The other big reason for using no-tracking is to get a view of the values in the database for entities that are already tracked by the context. In this case, fixing up to entities that are tracked by the context would not be desirable. However, doing one-way fixup to tracked entities (but not back the other way) for some forms of no-tracking query is an interesting idea that we will consider.

**Comment by AlirezaHaghshenas** [comment]
@alexk8 Fixing no-tracking entities to existing ones will result in strange behavior.

Consider this case: For simplicity let's say there is a one-to-one relationship between `Person` and `Position`:

You already have a `Person p` attached to the context.
You fetch `Job j1` as no tracking, its `Person` property will (as you demand) refer to p.
Then you fetch `Job j2` from the context. `Person p`'s `Job` property now refers to `Job j2`.

Now anyone using `Job j1` might fall victim to the contradiction that `j1.Person` refers to `p` but `p.Job` refers to `j2`

**Comment by ajcvickers** [comment]
@AlirezaHaghshenas If we did this the fixup would be one way only. That is the untracked objects would reference the tracked ones, but not vice versa. They would likely also be read-only. Essentially, this is a slight variation on the reference data scenario where some set of readonly entities don't need to be loaded every time they are used.

**Comment by AndriySvyryd** [comment]
Related to https://github.com/aspnet/EntityFrameworkCore/issues/10632
